File: 0003-longest-substring-without-repeating-characters/0003-longest-substring-without-repeating-characters.py
class Solution:
    def lengthOfLongestSubstring(self, s: str) -> int:
        length = 0
        substring = set()
        left = 0
        
        for right in range(len(s)):
            while s[right] in substring:
                substring.remove(s[left])
                left += 1
                    
            substring.add(s[right])
            length = max(length, right - left + 1)
            
        return length

File: 0003-longest-substring-without-repeating-characters/0003-longest-substring-without-repeating-characters.java
class Solution {
    public int lengthOfLongestSubstring(String s) {
        HashSet<Character> substring = new HashSet();
        int left = 0;
        int length = 0;
            
        for (int right = 0; right < s.length(); right++){
            while(substring.contains(s.charAt(right))){
                substring.remove(s.charAt(left));
                left++;
            }
            substring.add(s.charAt(right));
            length = Math.max(length, right - left + 1);
        }
        return length;
    }
}